File: tutorial/T17-MessagePubSub.cpp
/***************************
@Author: Chunel
@Contact: chunel@foxmail.com
@File: T17-Message-PubSub.cpp
@Time: 2022/12/27 00:19
@Desc: 本例子主要展示，通过 GMessageParam 在pipeline之间通过sub pub的方式传递
***************************/

#include "MyGNode/MyNode1.h"
#include "MyGNode/MyNode2.h"
#include "MyGNode/MyPubMessageNode.h"
#include "MyGNode/MySubMessageNode.h"

using namespace CGraph;

void sub_message() {
    CStatus status;
    GElementPtr node1, pubNode;
    GPipelinePtr pubPipeLine = GPipelineFactory::create();

    // 注册用于 发布 的pipeline
    status += pubPipeLine->registerGElement<MyNode1>(&node1, {}, "myNode1");
    status += pubPipeLine->registerGElement<MyPubMessageNode>(&pubNode, {});

    pubPipeLine->process(5);
}


void pub_message() {
    CStatus status;
    GElementPtr node2;
    GTemplateNodePtr<int> subNode;
    GPipelinePtr subPipeline = GPipelineFactory::create();
    // 每次调用bind方法的时候，会返回不同的 connId信息
    int connId = CGRAPH_BIND_MESSAGE_TOPIC(MyMessageParam, "pub-sub", 1024)

    // 注册用于 订阅 的pipeline
    status += subPipeline->registerGElement<MyNode2>(&node2, {}, "myNode2");
    status += subPipeline->registerGElement<MySubMessageNode<int>>(&subNode, {node2}, connId);
    subNode->setName("MySubMessageNode");

    subPipeline->process(5);
}


void tutorial_message_pub_sub() {
    // 一个pipeline pub数据，其他pipeline sub相同的数据内容
    std::thread pubThd = std::thread(sub_message);
    std::thread subThd1 = std::thread(pub_message);
    std::thread subThd2 = std::thread(pub_message);
    std::thread subThd3 = std::thread(pub_message);

    pubThd.join();
    subThd1.join();
    subThd2.join();
    subThd3.join();

    CGRAPH_CLEAR_MESSAGES()
    GPipelineFactory::clear();
}


int main() {
    tutorial_message_pub_sub();
    return 0;
}

File: python/tutorial/T17-MessagePubSub.py
"""
@Author: Chunel
@Contact: chunel@foxmail.com
@File: T17-MessagePubSub
@Time: 2026/7/26 15:15
@Desc: 一个 publisher pipeline 向三个 subscriber pipeline 发布消息

publish 只向每个 subscriber queue 复制同一个 Python 对象引用。发布后，
publisher 和 subscriber 都应把 payload 当作只读对象。

阻塞的 message 操作会占用当前 CGraph worker thread。在线程池过小，
或 DAG 依赖要求对端必须等当前节点完成后才能运行时，
WAIT 可能造成业务死锁。
"""

import threading

from pycgraph import GMessage, GPipeline

from MyGNode.MyNode1 import MyNode1
from MyGNode.MyNode2 import MyNode2
from MyGNode.MyPubMessageNode import MyPubMessageNode
from MyGNode.MySubMessageNode import MySubMessageNode


def pub_message():
    pipeline = GPipeline()
    node1 = MyNode1()
    pub_node = MyPubMessageNode()

    pipeline.registerGElement(node1, set(), "myNode1")
    pipeline.registerGElement(pub_node, set(), "MyPubMessageNode")
    return pipeline.process(5)


def sub_message(conn_ids):
    conn_id = GMessage.bind_message_topic("pub-sub", capacity=1024)
    conn_ids.append(conn_id)

    pipeline = GPipeline()
    node2 = MyNode2()
    sub_node = MySubMessageNode(conn_id)

    pipeline.registerGElement(node2, set(), "myNode2")
    pipeline.registerGElement(sub_node, {node2}, "MySubMessageNode")
    return pipeline.process(5)


def _run_concurrently(functions):
    statuses = [None] * len(functions)
    errors = []

    def execute(index, function):
        try:
            statuses[index] = function()
        except BaseException as error:
            errors.append(error)

    threads = [
        threading.Thread(target=execute, args=(index, function))
        for index, function in enumerate(functions)
    ]
    for thread in threads:
        thread.start()
    for thread in threads:
        thread.join()

    if errors:
        raise errors[0]
    for status in statuses:
        if not status.isOK():
            raise RuntimeError(status.getInfo())


def tutorial_message_pub_sub():
    conn_ids = []

    def subscriber():
        return sub_message(conn_ids)

    try:
        _run_concurrently(
            (pub_message, subscriber, subscriber, subscriber)
        )

        try:
            GMessage.sub_message(conn_ids[0], timeout_ms=10)
        except GMessage.Error as error:
            print("handled timeout:", error)
    finally:
        GMessage.clear_messages()


if __name__ == "__main__":
    tutorial_message_pub_sub()
